pkg/controller/acmeorders/sync.go

```diff
@@ -163,8 +172,30 @@ func (c *controller) Sync(ctx context.Context, o *cmacme.Order) (err error) {
 			}
 		}
 		return err
+
 	// anyChallengesFailed(challenges) == false is already implied by the above
-	// case, but explicitly check it here in case anything changes in future.
+	// case, but explicitly check it in the following cases for if anything changes in future.
+
+	// This is to avoid stuck Orders in edge cases where all the Challenges have
+	// been finalized, but the ACME server has not yet updated the ACME Order's
+	// status to valid. This is not an expected behaviour from an ACME server
+	// https://tools.ietf.org/html/rfc8555#section-7.1.6
+	// https://github.com/jetstack/cert-manager/issues/2868
+	case !anyChallengesFailed(challenges) && allChallengesFinal(challenges) && acmeOrder.Status == acmeapi.StatusPending:
+		log.V(logf.InfoLevel).Info("All challenges in a final state, waiting for ACME server to update the status of the order...")
+		// This is probably not needed as at this point the Order's status
+		// should already be Pending, but set it anyway to be explicit.
+		c.setOrderState(&o.Status, string(cmacme.Pending))
+		// Re-queue the Order to be processed after RequeuePeriod
+		key, err := cache.MetaNamespaceKeyFunc(o)
+		if err != nil {
+			// if we return an error here, at least the Order will get re-queued.
+			log.V(logf.InfoLevel).Info("Internal error: failed to construct key for pending Order")
+			return err
+		}
```

Comment:
I don't think it will get re-queued in this case. Presumably the same cache.MetaNamespaceKeyFunc is used in the callback functions of the informers so this error will have been encountered there and will never be encountered here.

So, I'd be inclined to log this as an unexpected runtime error here and return nil....but I may be missing something....and I don't suppose it'll ever happen anyway. Up to you.

Reply:
> Presumably the same cache.MetaNamespaceKeyFunc is used in the callback functions of the informers so this error will have been encountered there and will never be encountered here

You're right that it's used in callbacks already (we seem to use [this `KeyFunc`](https://github.com/jetstack/cert-manager/blob/master/pkg/controller/util.go#L34) which ends up calling the same `cache.MetaNamespaceKeyFunc`).
Perhaps there can be some weird edge case i.e with the `Order` object being somehow different when `cache.MetaNamespaceKeyFunc` is called here, but agree that that's likely not fixable with re-queue.
Not sure, if `Order` should be marked as failed in this scenario.
Perhaps I will just return a `nil` for now as it is very unlikely that this case would ever be encountered as you say.